37 improve coverage for initial classes

eriktamsen opened the pull request:
improved the test for the material and the setups to 100%
changed the definition of default values from the dictionary to the classes

pdiercks commented on fenicsxconcrete/experimental_setup/base_experiment.py:
```diff
@@ -22,36 +23,43 @@ def __init__(self, parameters: dict[str, pint.Quantity]):
+           """
 
-    def create_force_boundary(self, v):
+    def create_force_boundary(self, v = None):
         # define empty force boundary
         # TODO: is there a better solution???
 
         return None
```
hmm, the `BoundaryConditions` class is supposed to be able to handle this.
Do we want to use it here?

eriktamsen replied:
@pdiercks: I would suggest to use #29 for this, or should we focus on the displacement boundaries here and open a new issue for the force boundary implementation?

pdiercks replied:
yes, #29 is probably the better place to deal with that.

pdiercks commented on tests/finite_element_problem/test_log_levels.py:
```diff
@@ -0,0 +1,28 @@
+    log_str = log_level[0]
+    error_flag = log_level[1]
+
+    default_experiment, fem_parameters = LinearElasticity.default_parameters()
+    fem_parameters['log_level'] = log_str * ureg('')
```
would it make sense to have a method `set_log_level` for the base material?
I guess people might also expect to set the log level via integer values? (which does not seem to be possible?)

pdiercks replied:
this is also related to #45 or rather the discussion on what kind of logging should be used and then also how the user would control that.
set the log level for all classes at once or specify different log levels for experiment or material classes.

eriktamsen replied:
I agree, this should be dealt with in #45, this current implementation is be no means finished, optimized or ideal.
I think a single log level for everything would be best.

pdiercks commented on tests/test_helpers.py:
```diff
@@ -18,4 +18,10 @@ def test_parameters():
+    )
+
+
+def test_parameter_dic_functions():
+    parameters = Parameters()
+    # testing if adding None to dictionary works
+    parameters + None
```
```suggestion
def test_parameter_dic_functions():
    parameters = Parameters()
    # testing if adding None to dictionary works
    new = parameters + None
    assert new is parameters
```
?

pdiercks reviewed (commented):
Hi @eriktamsen ,
since you asked for it and I was bored on the train, here are some comments and suggestions and questions.

As far as I can tell after this we get 100% coverage for all files under `finite_element_problem` and `experimental_setup`.
I think that is very good.

only `base_sensor` and `other_sensor` are not covered 100%.
Did you plan to cover those in this PR as well?
In the issue it says setups and materials, but there is a new file in this PR that tests the sensor options: `tests/finite_element_problem/test_sensor_options.py`.
Should this be rather placed under `tests/sensors` (or so) instead?

Many of the comments are also kind of general and regarding the design.
If you feel that is not too relevant to the current aim of the PR feel free to discard that .

eriktamsen commented on fenicsxconcrete/finite_element_problem/base_material.py:
```diff
@@ -96,18 +99,14 @@ def __init__(self,
-        pass
+        """Must be defined by child"""
 
     def add_sensor(self, sensor):
         self.sensors[sensor.name] = sensor
```
I don't see this being a problem, but I added this check and a respective test.

eriktamsen commented on fenicsxconcrete/finite_element_problem/linear_elasticity.py:
```diff
@@ -22,7 +22,7 @@ def __init__(self,
         # adding default material parameter, will be overridden by outside input
         default_p = Parameters()
-        default_p['dummy'] = 42 * ureg('')   # just an example, maybe useful for other material classes?
+        default_p['dummy'] = 'example' * ureg('')  # example default parameter for this class
```
This is not a user input but a placeholder for a developer.
I just changed it to a line comment.

eriktamsen commented:
@pdiercks: thank for the review. I implemented your suggestions.
I did intentionally focus only on the coverage of the setup and the material.
The boundary conditions are taken care of in #42.
The sensor class needs to be improved first. There is an issue ready #43.
Regarding the question if the "test sensor options" should be moved, I would disagree, as this tests the functions of the material class, not the functions of the sensor class. I moved the test to a "test_base_material". Here we can test different functions of the base material that are inherited.

pdiercks reviewed (approved):
Thanks @eriktamsen !
Looks good to me
